### ISSUE TITLE
associative scan is incorrect for certain shapes/kwargs

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#137943]
I got this repro from noticing some test failures locally (details here: https://github.com/pytorch/pytorch/pull/136670#issuecomment-2412515964)
```
import torch
from torch._higher_order_ops.associative_scan import associative_scan

def f(a):
    return associative_scan(lambda x, y: x + y, a, dim=1, reverse=True, combine_mode="generic")

a = torch.arange(18, dtype=torch.float32, device='cuda').reshape(2, 9)

# I got this output from running with `_fake_associative_scan` defined here:
# https://github.com/pytorch/pytorch/blob/main/test/functorch/test_control_flow.py#L1579
expected_out = torch.tensor([
    [ 36.,  36.,  35.,  33.,  30.,  26.,  21.,  15.,   8.],
    [117., 108.,  98.,  87.,  75.,  62.,  48.,  33.,  17.],
], dtype=torch.float32, device='cuda')

out = f(a)
print(torch.allclose(out, expected_out))
# output is:
# tensor([[ 36.,  15.,  35.,   0.,  30.,   0.,  21.,   0.,   8.],
          [117.,  33.,  98.,  36.,  75.,  33.,  48.,  26.,  17.]],
         device='cuda:0')
```

**Comment by bdhirsh** [comment]
tentatively marking hi-pri for (what I think is?) silent correctness

**Comment by ydwu4** [comment]
Seems related to view operation. Need to take a further look at it.

**Comment by ydwu4** [comment]
@bohnstingl can take a look? This's the "generic" backend.

**Comment by bohnstingl** [comment]
Thank you @bdhirsh for raising this.
I think this is very much related to the issue that I already raised a long time ago.
https://github.com/pytorch/pytorch/issues/131805
Back then what I found is that this only happens for the scan dim being `9` and when the flip operation is involved, e.g., `reverse=True`.

Unfortunately, I tried to follow-up with @yf225 on this, but the issue is still open.
Could you try to see if in your case you also get a `Bad accuracy` warning at some point?

**Comment by bohnstingl** [comment]
So then i think it appears to be the same issue that I already reported.

> and nope, I didn't see any warnings while the test ran.

I only saw this warning when using `TORCH_LOGS=output_code,+dynamo`

**Comment by Chillee** [comment]
I will also take a look at this.

**Comment by bohnstingl** [comment]
Thank you @Chillee, I think this issue still persists and I don't quite know how to resolve it. I made a dedicated testcase that demonstrates this behavior: https://github.com/pytorch/pytorch/blob/8e9873d0a33267030df231240ed7e4b3929df251/test/functorch/test_control_flow.py#L2698

**Comment by ydwu4** [comment]
Minimal repro:
```
from torch._higher_order_ops.associative_scan import associative_scan
x = torch.ones(9, 1, device=torch.device("cuda"))
def f(x):
    return associative_scan(lambda a, b: a + b, x, dim=0, combine_mode="generic", reverse=True)

out = torch.compile(f, fullgraph=True)(x)
print(out)
```

This produces:
```
tensor([[9.],
        [2.],
        [7.],
        [0.],
        [5.],
        [0.],
        [3.],
        [0.],
        [1.]], device='cuda:0')
```
But  setting backend="eager" or "aot_eager" gives us the correct result:
```
tensor([[9.],
        [8.],
        [7.],
        [6.],
        [5.],
        [4.],
        [3.],
        [2.],
        [1.]], device='cuda:0')
```
The output code looks like: https://gist.github.com/ydwu4/a283086b9bf755d9d997628c0b4ff042

**Comment by ydwu4** [comment]
Running the bisector with following code:
```python
from torch._higher_order_ops.associative_scan import associative_scan
def f():
    torch._dynamo.reset()
    x = torch.ones(9, 1, device=torch.device("cuda"), dtype=torch.float32)
    out = associative_scan(lambda a, b: a + b, x, dim=0, combine_mode="generic", reverse=True)
    return torch.equal(out, torch.tensor([[9.], [8.], [7.], [6.], [5.], [4.], [3.], [2.], [1.],], dtype=torch.float32, device=torch.device("cuda")))

from torch._inductor.compiler_bisector import CompilerBisector
print(CompilerBisector.do_bisect(f))
```
 gives us:
```
No bisection status found.
Starting bisection process with system: eager
Moving to the next system: aot_eager
Moving to the next system: aot_eager_decomp_partition
Moving to the next system: aot_eager_decomp_partition_crossref
Moving to the next system: inductor
The issue is in the inductor system. Moving to the first subsystem: BisectSubsystem(name='joint_graph_passes')
Disabling joint_graph_passes did not fix the issue.
Moving to the next subsystem: inductor - post_grad_passes
Disabling post_grad_passes did not fix the issue.
Moving to the next subsystem: inductor - inductor_fallback_random
Disabling inductor_fallback_random did not fix the issue.
Moving to the next subsystem: inductor - inductor_emulate_precision_casts
Disabling inductor_emulate_precision_casts did not fix the issue.
Moving to the next subsystem: inductor - lowerings
Disabling lowerings fixed the issue.
Starting bisect by getting upper bound.
Upper bound of 33 found for inductor.
Bisecting inductor - lowerings (Range: [0, 33], Midpoint: 16)
Bisecting inductor - lowerings (Range: [17, 33], Midpoint: 25)
Bisecting inductor - lowerings (Range: [26, 33], Midpoint: 29)
Bisecting inductor - lowerings (Range: [30, 33], Midpoint: 31)
Bisecting inductor - lowerings (Range: [32, 33], Midpoint: 32)
Binary search completed for inductor - lowerings. The bisect number is 32. Debug info: rev_1
Bisection status deleted.
BisectionResult(backend='inductor', subsystem='lowerings', bisect_number=32, debug_info='rev_1')
```
This seems show that `rev_1` is the root cause.

**Comment by bohnstingl** [comment]
@ydwu4 I opened a PR for the compile backend change to "aot_eager" here https://github.com/pytorch/pytorch/pull/146973. However, this still does not resolve the problem and your repro script yields the wrong result. Only when changing this line
`out = torch.compile(f, fullgraph=True)(x)` to this line
`out = torch.compile(f, backend="aot_eager", fullgraph=True)(x)`
it works correctly.

**Comment by bohnstingl** [comment]
@ydwu4 this [PR](https://github.com/pytorch/pytorch/pull/146973) has landed in main, can you maybe try whether this issue still persists?

**Comment by ydwu4** [comment]
In the torch.compile(backend="inductor") path, this error would still persist i think because it overrides the "eager" backend decision.

**Comment by ydwu4** [comment]
Haven't got time to look into it lately.

**Comment by bohnstingl** [comment]
I just tried it on `main` and the issue still seems to persist.

**Comment by eellison** [comment]
No update yet

**Comment by ydwu4** [comment]
Will take a look this week.
